File: qt/multimedia/gen_qsoundeffect.go
package multimedia

import (
	"unsafe"

	"github.com/mappu/miqt/qt"
)

type QSoundEffect__Loop int

const (
	QSoundEffect__Infinite QSoundEffect__Loop = -2
)

type QSoundEffect__Status int

const (
	QSoundEffect__Null    QSoundEffect__Status = 0
	QSoundEffect__Loading QSoundEffect__Status = 1
	QSoundEffect__Ready   QSoundEffect__Status = 2
	QSoundEffect__Error   QSoundEffect__Status = 3
)

type QSoundEffect struct {
	h          uintptr
	isSubclass bool
}

// NewQSoundEffect constructs a new QSoundEffect object.
func NewQSoundEffect() *QSoundEffect {
	ret := newQSoundEffect(QSoundEffect_new())
	ret.isSubclass = true
	return ret
}

// NewQSoundEffect2 constructs a new QSoundEffect object.
func NewQSoundEffect2(audioDevice *QAudioDevice) *QSoundEffect {
	ret := newQSoundEffect(QSoundEffect_new2(audioDevice.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQSoundEffect3 constructs a new QSoundEffect object.
func NewQSoundEffect3(parent *qt.QObject) *QSoundEffect {
	ret := newQSoundEffect(QSoundEffect_new3((*QObject)(parent.UnsafePointer())))
	ret.isSubclass = true
	return ret
}

// NewQSoundEffect4 constructs a new QSoundEffect object.
func NewQSoundEffect4(audioDevice *QAudioDevice, parent *qt.QObject) *QSoundEffect {
	ret := newQSoundEffect(QSoundEffect_new4(audioDevice.cPointer(), (*QObject)(parent.UnsafePointer())))
	ret.isSubclass = true
	return ret
}

func (this *QSoundEffect) MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QSoundEffect_MetaObject(this.h)))
}

func (this *QSoundEffect) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QSoundEffect_Metacast(this.h, param1_Cstring))
}

func QSoundEffect_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QSoundEffect_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QSoundEffect_SupportedMimeTypes() []string {
	var _ma struct_miqt_array = QSoundEffect_SupportedMimeTypes()
	_ret := make([]string, int(_ma.len))
	_outCast := (*[0xffff]struct_miqt_string)(unsafe.Pointer(_ma.data)) // hey ya
	for i := 0; i < int(_ma.len); i++ {
		var _lv_ms struct_miqt_string = _outCast[i]
		_lv_ret := GoStringN(_lv_ms.data, int(int64(_lv_ms.len)))
		free(unsafe.Pointer(_lv_ms.data))
		_ret[i] = _lv_ret
	}
	return _ret
}

func (this *QSoundEffect) Source() *qt.QUrl {
	_goptr := qt.UnsafeNewQUrl(unsafe.Pointer(QSoundEffect_Source(this.h)))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QSoundEffect) SetSource(url *qt.QUrl) {
	QSoundEffect_SetSource(this.h, (*QUrl)(url.UnsafePointer()))
}

func (this *QSoundEffect) LoopCount() int {
	return (int)(QSoundEffect_LoopCount(this.h))
}

func (this *QSoundEffect) LoopsRemaining() int {
	return (int)(QSoundEffect_LoopsRemaining(this.h))
}

func (this *QSoundEffect) SetLoopCount(loopCount int) {
	QSoundEffect_SetLoopCount(this.h, (int)(loopCount))
}

func (this *QSoundEffect) AudioDevice() *QAudioDevice {
	_goptr := newQAudioDevice(QSoundEffect_AudioDevice(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QSoundEffect) SetAudioDevice(device *QAudioDevice) {
	QSoundEffect_SetAudioDevice(this.h, device.cPointer())
}

func (this *QSoundEffect) Volume() float32 {
	return (float32)(QSoundEffect_Volume(this.h))
}

func (this *QSoundEffect) SetVolume(volume float32) {
	QSoundEffect_SetVolume(this.h, (float)(volume))
}

func (this *QSoundEffect) IsMuted() bool {
	return (bool)(QSoundEffect_IsMuted(this.h))
}

func (this *QSoundEffect) SetMuted(muted bool) {
	QSoundEffect_SetMuted(this.h, (bool)(muted))
}

func (this *QSoundEffect) IsLoaded() bool {
	return (bool)(QSoundEffect_IsLoaded(this.h))
}

func (this *QSoundEffect) IsPlaying() bool {
	return (bool)(QSoundEffect_IsPlaying(this.h))
}

func (this *QSoundEffect) Status() Status {
	xxxxxxxxx
}

func (this *QSoundEffect) SourceChanged() {
	QSoundEffect_SourceChanged(this.h)
}

func (this *QSoundEffect) OnSourceChanged(slot func()) {
	QSoundEffect_connect_SourceChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QSoundEffect_SourceChanged
func miqt_exec_callback_QSoundEffect_SourceChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QSoundEffect) LoopCountChanged() {
	QSoundEffect_LoopCountChanged(this.h)
}

func (this *QSoundEffect) OnLoopCountChanged(slot func()) {
	QSoundEffect_connect_LoopCountChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QSoundEffect_LoopCountChanged
func miqt_exec_callback_QSoundEffect_LoopCountChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QSoundEffect) LoopsRemainingChanged() {
	QSoundEffect_LoopsRemainingChanged(this.h)
}

func (this *QSoundEffect) OnLoopsRemainingChanged(slot func()) {
	QSoundEffect_connect_LoopsRemainingChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QSoundEffect_LoopsRemainingChanged
func miqt_exec_callback_QSoundEffect_LoopsRemainingChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QSoundEffect) VolumeChanged() {
	QSoundEffect_VolumeChanged(this.h)
}

func (this *QSoundEffect) OnVolumeChanged(slot func()) {
	QSoundEffect_connect_VolumeChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QSoundEffect_VolumeChanged
func miqt_exec_callback_QSoundEffect_VolumeChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QSoundEffect) MutedChanged() {
	QSoundEffect_MutedChanged(this.h)
}

func (this *QSoundEffect) OnMutedChanged(slot func()) {
	QSoundEffect_connect_MutedChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QSoundEffect_MutedChanged
func miqt_exec_callback_QSoundEffect_MutedChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QSoundEffect) LoadedChanged() {
	QSoundEffect_LoadedChanged(this.h)
}

func (this *QSoundEffect) OnLoadedChanged(slot func()) {
	QSoundEffect_connect_LoadedChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QSoundEffect_LoadedChanged
func miqt_exec_callback_QSoundEffect_LoadedChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QSoundEffect) PlayingChanged() {
	QSoundEffect_PlayingChanged(this.h)
}

func (this *QSoundEffect) OnPlayingChanged(slot func()) {
	QSoundEffect_connect_PlayingChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QSoundEffect_PlayingChanged
func miqt_exec_callback_QSoundEffect_PlayingChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QSoundEffect) StatusChanged() {
	QSoundEffect_StatusChanged(this.h)
}

func (this *QSoundEffect) OnStatusChanged(slot func()) {
	QSoundEffect_connect_StatusChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QSoundEffect_StatusChanged
func miqt_exec_callback_QSoundEffect_StatusChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QSoundEffect) AudioDeviceChanged() {
	QSoundEffect_AudioDeviceChanged(this.h)
}

func (this *QSoundEffect) OnAudioDeviceChanged(slot func()) {
	QSoundEffect_connect_AudioDeviceChanged(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QSoundEffect_AudioDeviceChanged
func miqt_exec_callback_QSoundEffect_AudioDeviceChanged(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func (this *QSoundEffect) Play() {
	QSoundEffect_Play(this.h)
}

func (this *QSoundEffect) Stop() {
	QSoundEffect_Stop(this.h)
}

func QSoundEffect_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QSoundEffect_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QSoundEffect_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QSoundEffect_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QSoundEffect) callVirtualBase_MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QSoundEffect_virtualbase_MetaObject(unsafe.Pointer(this.h))))
}

func (this *QSoundEffect) OnMetaObject(slot func(super func() *qt.QMetaObject) *qt.QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QSoundEffect_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QSoundEffect_MetaObject
func miqt_exec_callback_QSoundEffect_MetaObject(self QSoundEffect, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *qt.QMetaObject) *qt.QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QSoundEffect{h: self}).callVirtualBase_MetaObject)

	return (*QMetaObject)(virtualReturn.UnsafePointer())
}

func (this *QSoundEffect) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QSoundEffect_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QSoundEffect) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QSoundEffect_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QSoundEffect_Metacast
func miqt_exec_callback_QSoundEffect_Metacast(self QSoundEffect, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QSoundEffect{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}

File: qt/multimedia/gen_qsoundeffect.cpp
// +build ignore

#include <QAudioDevice>
#include <QList>
#include <QMetaObject>
#include <QObject>
#include <QSoundEffect>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <QUrl>
#include <qsoundeffect.h>
#include "gen_qsoundeffect.h"
class MiqtVirtualQSoundEffect : public virtual QSoundEffect {
public:
MiqtVirtualQSoundEffect(): QSoundEffect() {};
MiqtVirtualQSoundEffect(const QAudioDevice& audioDevice): QSoundEffect(audioDevice) {};
MiqtVirtualQSoundEffect(QObject* parent): QSoundEffect(parent) {};
MiqtVirtualQSoundEffect(const QAudioDevice& audioDevice, QObject* parent): QSoundEffect(audioDevice, parent) {};

virtual ~MiqtVirtualQSoundEffect() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QSoundEffect::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QSoundEffect_MetaObject(const_cast<MiqtVirtualQSoundEffect*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QSoundEffect::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QSoundEffect::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QSoundEffect_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QSoundEffect::qt_metacast(param1);

	}
};
QSoundEffect* QSoundEffect_new() {
return new MiqtVirtualQSoundEffect();
}
QSoundEffect* QSoundEffect_new2(QAudioDevice* audioDevice) {
return new MiqtVirtualQSoundEffect(*audioDevice);
}
QSoundEffect* QSoundEffect_new3(QObject* parent) {
return new MiqtVirtualQSoundEffect(parent);
}
QSoundEffect* QSoundEffect_new4(QAudioDevice* audioDevice, QObject* parent) {
return new MiqtVirtualQSoundEffect(*audioDevice, parent);
}
void QSoundEffect_virtbase(QSoundEffect* src
, QObject** outptr_QObject
) {
*outptr_QObject = static_cast<QObject*>(src);
}
QMetaObject* QSoundEffect_MetaObject(const QSoundEffect* self) {
	return (QMetaObject*) self->metaObject();
}
void* QSoundEffect_Metacast(QSoundEffect* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QSoundEffect_Tr(const char* s) {
	QString _ret = QSoundEffect::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_array /* of struct miqt_string */  QSoundEffect_SupportedMimeTypes() {
	QStringList _ret = QSoundEffect::supportedMimeTypes();
	// Convert QList<> from C++ memory to manually-managed C memory
	struct miqt_string* _arr = static_cast<struct miqt_string*>(malloc(sizeof(struct miqt_string) * _ret.length()));
	for (size_t i = 0, e = _ret.length(); i < e; ++i) {
		QString _lv_ret = _ret[i];
		// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
		QByteArray _lv_b = _lv_ret.toUtf8();
		struct miqt_string _lv_ms;
		_lv_ms.len = _lv_b.length();
		_lv_ms.data = static_cast<char*>(malloc(_lv_ms.len));
		memcpy(_lv_ms.data, _lv_b.data(), _lv_ms.len);
		_arr[i] = _lv_ms;
	}
	struct miqt_array _out;
	_out.len = _ret.length();
	_out.data = static_cast<void*>(_arr);
	return _out;
}
QUrl* QSoundEffect_Source(const QSoundEffect* self) {
	return new QUrl(self->source());
}
void QSoundEffect_SetSource(QSoundEffect* self, QUrl* url) {
	self->setSource(*url);
}
int QSoundEffect_LoopCount(const QSoundEffect* self) {
	return self->loopCount();
}
int QSoundEffect_LoopsRemaining(const QSoundEffect* self) {
	return self->loopsRemaining();
}
void QSoundEffect_SetLoopCount(QSoundEffect* self, int loopCount) {
	self->setLoopCount(static_cast<int>(loopCount));
}
QAudioDevice* QSoundEffect_AudioDevice(QSoundEffect* self) {
	return new QAudioDevice(self->audioDevice());
}
void QSoundEffect_SetAudioDevice(QSoundEffect* self, QAudioDevice* device) {
	self->setAudioDevice(*device);
}
float QSoundEffect_Volume(const QSoundEffect* self) {
	return self->volume();
}
void QSoundEffect_SetVolume(QSoundEffect* self, float volume) {
	self->setVolume(static_cast<float>(volume));
}
bool QSoundEffect_IsMuted(const QSoundEffect* self) {
	return self->isMuted();
}
void QSoundEffect_SetMuted(QSoundEffect* self, bool muted) {
	self->setMuted(muted);
}
bool QSoundEffect_IsLoaded(const QSoundEffect* self) {
	return self->isLoaded();
}
bool QSoundEffect_IsPlaying(const QSoundEffect* self) {
	return self->isPlaying();
}
Status QSoundEffect_Status(const QSoundEffect* self) {
	return self->status();
}
void QSoundEffect_SourceChanged(QSoundEffect* self) {
	self->sourceChanged();
}
void QSoundEffect_connect_SourceChanged(QSoundEffect* self, intptr_t slot) {
	MiqtVirtualQSoundEffect::connect(self, static_cast<void (QSoundEffect::*)()>(&QSoundEffect::sourceChanged), self, [=]() {
		miqt_exec_callback_QSoundEffect_SourceChanged(slot);
	});
}
void QSoundEffect_LoopCountChanged(QSoundEffect* self) {
	self->loopCountChanged();
}
void QSoundEffect_connect_LoopCountChanged(QSoundEffect* self, intptr_t slot) {
	MiqtVirtualQSoundEffect::connect(self, static_cast<void (QSoundEffect::*)()>(&QSoundEffect::loopCountChanged), self, [=]() {
		miqt_exec_callback_QSoundEffect_LoopCountChanged(slot);
	});
}
void QSoundEffect_LoopsRemainingChanged(QSoundEffect* self) {
	self->loopsRemainingChanged();
}
void QSoundEffect_connect_LoopsRemainingChanged(QSoundEffect* self, intptr_t slot) {
	MiqtVirtualQSoundEffect::connect(self, static_cast<void (QSoundEffect::*)()>(&QSoundEffect::loopsRemainingChanged), self, [=]() {
		miqt_exec_callback_QSoundEffect_LoopsRemainingChanged(slot);
	});
}
void QSoundEffect_VolumeChanged(QSoundEffect* self) {
	self->volumeChanged();
}
void QSoundEffect_connect_VolumeChanged(QSoundEffect* self, intptr_t slot) {
	MiqtVirtualQSoundEffect::connect(self, static_cast<void (QSoundEffect::*)()>(&QSoundEffect::volumeChanged), self, [=]() {
		miqt_exec_callback_QSoundEffect_VolumeChanged(slot);
	});
}
void QSoundEffect_MutedChanged(QSoundEffect* self) {
	self->mutedChanged();
}
void QSoundEffect_connect_MutedChanged(QSoundEffect* self, intptr_t slot) {
	MiqtVirtualQSoundEffect::connect(self, static_cast<void (QSoundEffect::*)()>(&QSoundEffect::mutedChanged), self, [=]() {
		miqt_exec_callback_QSoundEffect_MutedChanged(slot);
	});
}
void QSoundEffect_LoadedChanged(QSoundEffect* self) {
	self->loadedChanged();
}
void QSoundEffect_connect_LoadedChanged(QSoundEffect* self, intptr_t slot) {
	MiqtVirtualQSoundEffect::connect(self, static_cast<void (QSoundEffect::*)()>(&QSoundEffect::loadedChanged), self, [=]() {
		miqt_exec_callback_QSoundEffect_LoadedChanged(slot);
	});
}
void QSoundEffect_PlayingChanged(QSoundEffect* self) {
	self->playingChanged();
}
void QSoundEffect_connect_PlayingChanged(QSoundEffect* self, intptr_t slot) {
	MiqtVirtualQSoundEffect::connect(self, static_cast<void (QSoundEffect::*)()>(&QSoundEffect::playingChanged), self, [=]() {
		miqt_exec_callback_QSoundEffect_PlayingChanged(slot);
	});
}
void QSoundEffect_StatusChanged(QSoundEffect* self) {
	self->statusChanged();
}
void QSoundEffect_connect_StatusChanged(QSoundEffect* self, intptr_t slot) {
	MiqtVirtualQSoundEffect::connect(self, static_cast<void (QSoundEffect::*)()>(&QSoundEffect::statusChanged), self, [=]() {
		miqt_exec_callback_QSoundEffect_StatusChanged(slot);
	});
}
void QSoundEffect_AudioDeviceChanged(QSoundEffect* self) {
	self->audioDeviceChanged();
}
void QSoundEffect_connect_AudioDeviceChanged(QSoundEffect* self, intptr_t slot) {
	MiqtVirtualQSoundEffect::connect(self, static_cast<void (QSoundEffect::*)()>(&QSoundEffect::audioDeviceChanged), self, [=]() {
		miqt_exec_callback_QSoundEffect_AudioDeviceChanged(slot);
	});
}
void QSoundEffect_Play(QSoundEffect* self) {
	self->play();
}
void QSoundEffect_Stop(QSoundEffect* self) {
	self->stop();
}
struct miqt_string QSoundEffect_Tr2(const char* s, const char* c) {
	QString _ret = QSoundEffect::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QSoundEffect_Tr3(const char* s, const char* c, int n) {
	QString _ret = QSoundEffect::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QSoundEffect_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQSoundEffect*>( (QSoundEffect*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QSoundEffect_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQSoundEffect*)(self) )->virtualbase_MetaObject();
}
void QSoundEffect_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQSoundEffect*>( (QSoundEffect*)(self) )->handle__Metacast = slot;
}
void* QSoundEffect_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQSoundEffect*)(self) )->virtualbase_Metacast(param1);
}
void QSoundEffect_Delete(QSoundEffect* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQSoundEffect*>( self );
	} else {
		delete self;
	}
}
